File: src/vm/mod/iodev/test/jumpers.cpp
#include "jumpers.hpp"

namespace kcpu::iodev {

jumpers::jumpers(ctl_out_interface &ctl, pic_in_interface &pic) : single_port_io_device(PORT_BASE), ctl(ctl), pic(pic) {
}

std::pair<regval_t, halfcycle_count_t> jumpers::read() {
    return std::pair(flags, 0);
}

halfcycle_count_t jumpers::write(regval_t val) {
    flags = val;
    return 0;
}

void jumpers::process_halfcycle(bool offclock) {
    if(flags & FLAG_TUI2NMI) {
        if(offclock && ctl.is_tui_active()) {
            pic.assert(pic_in_interface::NMI_INT);
        }
    }
}

}

File: src/vm/io/dev/test/jumpers.rs
use super::super::pic::Pic;
use crate::spec::types::hw::*;
use crate::vm::interface;
use crate::vm::io::types::*;

const PORT_BASE: Word = 0xD0;
// Connect the TUI line of CTL, ANDed with NOT_CLOCK, to the NMI assert of the PIC
const FLAG_TUI2NMI: Word = 0x0001;

pub struct Jumpers {
    pic: Handle<Pic>,

    flags: Word,
}

impl Jumpers {
    pub fn new(pic: Handle<Pic>) -> Self {
        Self { pic, flags: 0 }
    }
}

impl SinglePortDevice for Jumpers {
    fn reserved_port(&self) -> Word {
        PORT_BASE
    }

    fn write(&mut self, val: Word) -> HalfcycleCount {
        self.flags = val;
        0
    }

    fn read(&mut self) -> (HalfcycleCount, Word) {
        (0, self.flags)
    }

    fn process_halfcycle(&mut self, sigs: ClockedSignals) {
        if self.flags & FLAG_TUI2NMI != 0 {
            if let ClockedSignals::OffClock(_, true) = sigs {
                interface::Pic::assert(&self.pic, interface::PicIrq::Nmi);
            }
        }
    }
}
